text/0000-glimmer-components.md

```diff
@@ -0,0 +1,1019 @@
+
+## Prior Art
+
+Part of the challenge in the original Angle Bracket components RFC was
+attempting to design without implementation, testing, usage, and feedback.
+Glimmer.js provided an early method to experiment, but because it was not widely
+adopted there wasn't much feedback from larger-scale usage. This in part
+motivated the [component manager
+RFC](https://github.com/emberjs/rfcs/blob/master/text/0213-custom-components.md),
+which enabled experimentation in Ember directly, and set us up for having
+multiple implementations of component APIs which were interchangeable.
+
+As such, we now have two reference implementations which can be referred to:
+
+* [Glimmer.js](https://glimmerjs.com/), the framework that this component API is
+  based on, and will be cross-compatible with.
+* [sparkles-component](https://github.com/rwjblue/sparkles-component), an an
+  implementation of the Glimmer.js component API using Ember's component
+  managers. It is usable in Ember today.
+
+Both of these have minor differences from the API proposed in this RFC, mainly
+because they were made before the [element modifiers
+RFC](https://github.com/emberjs/rfcs/blob/master/text/0373-Element-Modifier-Managers.md)
+was accepted and opened up additional design possibilities. However, they serve
+as valuable data points for the viability of a simpler component API, and inform
+the design accordingly.
+
+## Detailed design
+
+Glimmer components have the following interface:
+
+```ts
+interface GlimmerComponent<T = object> {
+  args: T;
+
+  isDestroying: boolean;
+  isDestroyed: boolean;
+
+  constructor(owner: Owner, args: T): void;
+
+  didUpdateArgs(): void;
+  willDestroy(): void;
+}
+```
+
+In addition, two new element modifiers will fill in the gaps left by the
+`didInsertElement`, `didRender`, and `willDestroyElement` lifecycle hooks:
+
+* `{{did-render}}`
+* `{{will-destroy}}`
+
+### Constructor
+
+The constructor for Glimmer components receives two arguments: The owner
+instance and the named arguments object. Both of these arguments should
+conventionally be passed to `super` immediately, and then accessed through
+decorated service properties, `getOwner`, and `this.args`:
+
+```js
+class PersonComponent extends GlimmerComponent {
+  @service profile;
+
+  constructor() {
+    super(...arguments);
+
+    let owner = getOwner(this);
+    let profileService = this.profile;
+
+    let firstName = this.args.firstName;
+  }
+}
+```
+
+These arguments are passed to the constructor so that they can be used for
+initial setup of the component. Service injections and args being available in
+this way also makes them available to class field initializers, which can be
+useful:
+
+```js
+class PersonComponent extends GlimmerComponent {
+  @service time;
+
+  // Use the values of args in an initializer
+  fullName = `${this.args.firstName} ${this.args.lastName}`;
+
+  // Access a service in an initializer
+  currentTime = this.time.now();
```

Comment:
Initialized values are available after the call to `super()` in the constructor, if I'm reading the first part of what you're saying correctly, so that should not be an issue.

We should absolutely touch on the nuances of classes in the learning model, I will call this out in the "How we teach this" section.